add maxRestarts to RestartWithBackoff #24129

Sebruck opened the pull request:
Implements #24129

The new option maxRestarts will limit the amount of restarts a RestartSource, -Flow or -Sink is allowed to do.

Commit:
add maxRestarts to RestartWithBackoff #24129

The new option maxRestarts will limit the amount of restarts a
RestartSource, -Flow or -Sink is allowed to do.

Sebruck commented on akka-stream/src/main/scala/akka/stream/scaladsl/Restart.scala:
```diff
@@ -37,10 +37,11 @@ object RestartSource {
    */
-  def withBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double)(sourceFactory: () ⇒ Source[T, _]): Source[T, NotUsed] = {
-    Source.fromGraph(new RestartWithBackoffSource(sourceFactory, minBackoff, maxBackoff, randomFactor, onlyOnFailures = false))
+  def withBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double, maxRestarts: Int)(sourceFactory: () ⇒ Source[T, _]): Source[T, NotUsed] = {
```
Having `maxRestarts: Int` would require to cap the amount of restarts always. I am not sure if that's a good idea. Maybe it's better to have it as an `Option[Int]` where `None` would mean that amount of restarts are not capped.

Anyway this would be a breaking change to the API, or? Should there be a default parameter (like `None`)?

Depending on what we discuss, how should the java api look like?

@hseeberger what do you think?

hseeberger replied:
So far Akka is rather using "magic numbers" (-1, Int.MaxValue, etc.) instead of `Option`s. Which is probabaly due to the lack of `Optional` prior to Java 8. Hence I'm fine with an `Int` instead of an `Option[Int]`, because it's consistent(ly bad).

Sebruck replied:
Okay, I will add some logic that no limiting is applied when `maxRestarts <= 0` (since having a RestartSource with 0 restarts does not make sense at all).

What about my other question:
Should I add a default value, e.g. `maxRestarts: Int = -1` to not break existing usages?

hseeberger commented on akka-stream/src/main/scala/akka/stream/javadsl/Restart.scala:
```diff
@@ -36,10 +36,11 @@ object RestartSource {
    */
-  def withBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double, sourceFactory: Creator[Source[T, _]]): Source[T, NotUsed] = {
-    akka.stream.scaladsl.RestartSource.withBackoff(minBackoff, maxBackoff, randomFactor) { () ⇒
+  def withBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double, maxRestarts: Int, sourceFactory: Creator[Source[T, _]]): Source[T, NotUsed] = {
```
As we have no default value in Java, maybe provide an overloaded alternative which is using `Int.MaxValue`?

hseeberger commented on akka-stream/src/main/scala/akka/stream/scaladsl/Restart.scala:
```diff
@@ -348,14 +356,17 @@ private abstract class RestartWithBackoffLogic[S <: Shape](
 
-  // Set a timer to restart after the calculated delay
-  protected final def scheduleRestartTimer() = {
+  protected final def maxRestartsReached() = {
```
Style protip: No need for the empty parameter list here.

Sebruck replied:
Thanks for the feedback.

I decided explicitly to have them here to give a hint that this function has a side effect (resetting the restartCount). I would find it even more misleading if the parameter list is omitted.

@hseeberger Are you fine with that?

hseeberger replied:
It doesn't have a side-effect, it's a mere getter (returning a proper value), right? 
Getter-like methods should be without parantheses.

Sebruck replied:
Since it mutates a variable outside of its scope (`restartCount`), I was considering this function as "side-effectful".

Anyways I have no strong opinions on this and I trust your judgments, I just removed the parameter list.

Sebruck commented:
I guess I need to update some documentation, or? Can anyone give me a hint where to start?

Commit:
consider maxRestarts smaller than 1 as no restart limit

Sebruck commented:
I now consider maxRestarts smaller than 0 as no restarts and adapted the java api. I like the "-1" more than Int.MaxValue, since Int.MaxValue could theoretically be reached.

hseeberger commented:
Please keep MaxValue, this is how the other APIs are working, e.g. Akka Persistence Query.

Sebruck commented:
Ok sorry, didn't know that, I'll change that. Thought that I can choose since you offered both :-)

Commit:
Use Int.MaxValue as default for maxRestarts instead of -1

* removed the special treatment of negative numbers

Commit:
Remove parameter list of maxRestartsReached

hseeberger commented:
LGTM  👍

Commit:
Add maxRestarts to documentation

johanandren commented:
OK TO TEST

johanandren commented on akka-stream/src/main/scala/akka/stream/javadsl/Restart.scala:
```diff
@@ -38,12 +38,41 @@ object RestartSource {
+   * is always handled by restarting it. The wrapped [[Source]] can however be cancelled by cancelling this [[Source]].
+   * When that happens, the wrapped [[Source]], if currently running will be cancelled, and it will not be restarted.
+   * This can be triggered simply by the downstream cancelling, or externally by introducing a [[KillSwitch]] right
+   * after this [[Source]] in the graph.
```
This is not quite true anymore with the `maxRestarts` right?

Sebruck replied:
Improved the documentation of the Source Sink and Flow

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Restart.scala:
```diff
@@ -348,14 +356,17 @@ private abstract class RestartWithBackoffLogic[S <: Shape](
 
-  // Set a timer to restart after the calculated delay
-  protected final def scheduleRestartTimer() = {
+  protected final def maxRestartsReached = {
```
As it also side effects, please add an empty parameter list to the method

hseeberger replied:
@johanandren is the ultimate authority here, I'm just Mr Style ;-) So let there be parenthesis in order to proceed.

johanandren replied:
Rule of thumb is empty parameter list to hint that something isn't pure. Makes things a bit easier to reason about.

hseeberger replied:
What’s pure?

johanandren replied:
As in not doing any side effects.

Sebruck replied:
✅ added the empty parameter list

johanandren commented on akka-stream/src/main/scala/akka/stream/javadsl/Restart.scala:
```diff
@@ -38,12 +38,41 @@ object RestartSource {
+   * @param randomFactor after calculation of the exponential back-off an additional
+   *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
+   *   In order to skip this additional delay pass in `0`.
+   * @param maxRestarts the amount of restarts is capped to this amount.
```
I think it also needs an explanation that the maxRestarts isn't "global" but within a timespan (and what timespan that is)

Sebruck replied:
Improved the documentation, can you please verify if it's better now?

johanandren reviewed (commented):
Added some minor comments, it also needs some mima filters as it adds new methods in public apis (look for the `ProblemFilters.exclude` in the failed CI build log)

Commit:
Improve documentation

Commit:
Add empty parameter list to make clear that the function is not pure

Commit:
Add new changed Restart Source/Sink/Flow to mima excludes

johanandren commented:
Unrelated failure.

johanandren commented on akka-stream/src/main/mima-filters/2.5.8.backwards.excludes:
```diff
@@ -45,4 +45,10 @@ ProblemFilters.exclude[MissingClassProblem]("akka.stream.stage.GraphStageLogic$C
-ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartWithBackoffSource.this")
\ No newline at end of file
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartWithBackoffSource.this")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartSource.onFailuresWithBackoff")
```
Please add an empty line with a comment referencing the issue and a minimal description (the above is one single such chunk).

Sebruck replied:
Added a comment, should I squash all the commits into the first one or is this done by the CI / the merger?

johanandren replied:
We can do that through the github ui when we merge, thanks!

johanandren reviewed (approved):
One more tiny thing with the mima-excludes file, then it is LGTM!

Commit:
Add comment to mima filters

Sebruck commented:
What's the procedure to get this merged?

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Restart.scala:
```diff
@@ -60,11 +63,12 @@ object RestartSource {
    */
-  def onFailuresWithBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double)(sourceFactory: () ⇒ Source[T, _]): Source[T, NotUsed] = {
-    Source.fromGraph(new RestartWithBackoffSource(sourceFactory, minBackoff, maxBackoff, randomFactor, onlyOnFailures = true))
+  def onFailuresWithBackoff[T](minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double, maxRestarts: Int = Int.MaxValue)(sourceFactory: () ⇒ Source[T, _]): Source[T, NotUsed] = {
```
-1, using default parameters will make this API impossible to evolve in a binary compatible way; please use overloaded methods same as in Java.

Sebruck replied:
✅

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Restart.scala:
```diff
@@ -144,9 +235,41 @@ object RestartFlow {
+   * @param randomFactor after calculation of the exponential back-off an additional
+   *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
+   *   In order to skip this additional delay pass in `0`.
+   * @param maxRestarts the amount of restarts is capped to this amount within a time frame of minBackoff.
```
Define in docs what a 0 or a negative number will mean

Sebruck replied:
✅

ktoso reviewed (changes requested):
Even open source developers need a break, and such special family time is during christmas/newyear 😉 Thanks for your patience, it's mostly about that, that the comitters just need a break for a week to launch with more energy into 2018 -- hope you can relate to that.

I did review nevertheless, thinking maybe it's ready, but I did find some minor things that have to be tweaked before this is ready. Please address those when you have the time, and see you in 2018, thanks!

Sebruck commented:
Thanks for your reply! My question was not meant to be pushy in any way, I was just wondering what's the usual procedure and if I need to take some action :-)

Commit:
Do not use default parameters in the public api

Default parameters will make the api impossible to evolve in a binary
compatible way

Commit:
Document the behaviour of maxRestarts for `0` and negative numbers

patriknw commented on akka-stream/src/main/mima-filters/2.5.8.backwards.excludes:
```diff
@@ -45,4 +45,11 @@ ProblemFilters.exclude[MissingClassProblem]("akka.stream.stage.GraphStageLogic$C
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartSink.withBackoff")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartWithBackoffSink.this")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartWithBackoffFlow.this")
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartFlow.withBackoff")
```
this is adding too many filters that are not needed. It should be enough with
```
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartWithBackoffSink.this")
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.RestartWithBackoffFlow.this")
```

apart from that this PR looks good and can be merged

Commit:
Reduce amount of mima filters

patriknw commented:
I have fixed the conflict. WIll be merged shortly in https://github.com/akka/akka/pull/24328. thanks!

Sebruck commented:
Thanks for all your support!
